ui/src/app/base/components/MainTable/MainTable.js

```diff
@@ -0,0 +1,209 @@
+          _updateSort.bind(
+            this,
+            _setSortKey,
+            _setSortDirection,
+            sortKey,
+            sortDirection
+          )
+        }
+        {...props}
+      >
+        {content}
+      </TableHeader>
+    );
+  });
+  // When there is expanding content then provide an extra hidden header to
+  // account for the extra cell in the body rows.
+  return (
+    <thead>
+      <TableRow>
+        {headerItems}
+        {expanding && <TableHeader className="u-hide"></TableHeader>}
+      </TableRow>
+    </thead>
+  );
+};
+const _generateRows = (
+  currentSortDirection,
+  currentSortKey,
+  expanding,
+  rows,
+  sortable
+) => {
+  // Clone the rows so we can restore the original order.
+  const sortedRows = [...rows];
+  if (sortable && currentSortKey) {
+    sortedRows.sort((a, b) => {
+      if (!a.sortData || !b.sortData) {
+        return 0;
+      }
+      if (a.sortData[currentSortKey] > b.sortData[currentSortKey]) {
+        return currentSortDirection === "ascending" ? 1 : -1;
+      } else if (a.sortData[currentSortKey] < b.sortData[currentSortKey]) {
+        return currentSortDirection === "ascending" ? -1 : 1;
+      }
+      return 0;
+    });
+  }
+  const rowItems = sortedRows.map(
+    ({ columns, expanded, expandedContent, sortData, ...rowProps }, index) => {
+      const cellItems = columns.map(({ content, ...cellProps }, index) => (
+        <TableCell key={index} {...cellProps}>
+          {content}
+        </TableCell>
+      ));
+      // The expanding cell is alway created to match the correct number of
+      // table cells in rows that do have expanding content.
+      return (
+        <TableRow key={index} {...rowProps}>
+          {cellItems}
+          {expanding && (
+            <TableCell expanding={true} hidden={!expanded}>
+              {expandedContent}
+            </TableCell>
+          )}
+        </TableRow>
+      );
+    }
+  );
+  return <tbody>{rowItems}</tbody>;
+};
+
+const MainTable = ({
+  defaultSort,
+  defaultSortDirection,
+  expanding,
+  headers,
+  rows,
+  responsive,
+  sortable,
+  ...props
+}) => {
+  const [currentSortKey, _setSortKey] = useState(defaultSort);
+  const [currentSortDirection, _setSortDirection] = useState(
+    defaultSortDirection
+  );
+
+  // Update the current sort state if the prop changes.
+  useEffect(() => {
+    _setSortKey(defaultSort);
+  }, [defaultSort]);
+
+  // Update the current sort direction state if the prop changes.
+  useEffect(() => {
+    _setSortDirection(defaultSortDirection);
+  }, [defaultSortDirection]);
+
+  return (
+    <Table
+      expanding={expanding}
+      sortable={sortable}
+      responsive={responsive}
+      {...props}
+    >
+      {headers &&
+        _generateHeaders(
+          currentSortKey,
+          currentSortDirection,
+          expanding,
+          headers,
+          sortable,
+          _setSortKey,
+          _setSortDirection
+        )}
+      {headers &&
```

Comment:
as discussed on irc, this should probably be `rows`

Reply:
Thanks.